compiler/rustc_middle/src/middle/stability.rs

```diff
@@ -292,6 +303,32 @@ fn skip_stability_check_due_to_privacy(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
     }
 }
 
+// See issue #83250.
+fn get_sugg_for_allocator_api(
+    tcx: TyCtxt<'_>,
+    def_id: DefId,
+    span: Span,
+    feature: Symbol,
+) -> Option<(Span, String, String, Applicability)> {
+    if feature == Symbol::intern("allocator_api") {
+        if let Some(trait_) = tcx.parent(def_id) {
+            if tcx.is_diagnostic_item(sym::Vec, trait_) {
```

Comment:
I was more thinking that you could add a new diagnostic item to the `A` generic parameter itself and then check if `def_id` had that diagnostic item, rather than check if `def_id`'s parent is a `Vec`. I'm not entirely sure that we *can* do that, might not be able to put the attribute on the parameter, but worth a try. If you added diagnostic items like that to any `allocator_api` generic parameter then you can just check if any of them match here, and we'd be providing a suggestion for non-`Vec` cases too.

Reply:
@davidtwco 
Thank you so much for the additional reviews :)

In my understanding, we have the `feature` variable there, so we can just put only the first if-condition in case we'd suggest all type cases. Accordingly, we can omit an additional diagnostic item and def_id's parent checks. Is it correct that we'd like to expand the Vec suggestion to any other types on `allocator_api`? Or is there another purpose in adding a new diagnostic item?

Reply:
@davidtwco 
I updated the commit with my understanding. Could you please take a look?

Reply:
I just think it would be neater to check that `def_id` is the specific allocator generic argument rather assume that if we have a stability error with `allocator_api` and the parent type is a `Vec` then it must be because of the allocator generic argument (in practice, this would probably be fine, but if we can be more precise then I think that would be preferable).

Either approach can be extended to other types where this sort of suggestion would make sense.

Reply:
@davidtwco 
Ah, I see. That indeed would be precise. I'd do some research on it. Thank you :)

Reply:
@davidtwco 
I've put a new diagnostic item on the [`Allocator` trait](https://github.com/rust-lang/rust/blob/934624fe5f66ce3fb8abf0597a6deb079783335f/library/core/src/alloc/mod.rs#L100) that indicates the allocator generic argument, but it seems types that don't have impl of the trait such as `_` won't have its item. In my view, this way cannot be used here.

Reply:
What if you put the attribute on the allocator generic argument of the `Vec` type?

https://github.com/rust-lang/rust/blob/faea820643f9acec1bb787941063393f9508a775/library/alloc/src/vec/mod.rs#L400

Reply:
@davidtwco 
That also shows the same behavior...
Just in case, could you confirm that the following way I've tried is correct?

`rust/library/alloc/src/vec/mod.rs`

```diff
- pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
+ pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] #[rustc_diagnostic_item = "VecAllocator"] A: Allocator = Global> {
```

`rust/compiler/rustc_span/src/symbol.rs`

```diff
  Vec,
+ VecAllocator,
  VecDeque,
```

`rust/compiler/rustc_middle/src/middle/stability.rs`

```diff
fn suggestion_for_allocator_api(
...
) -> Option<(Span, String, String, Applicability)> {
+    println!("type: {}", tcx.is_diagnostic_item(sym::VecAllocator, def_id));
+    if let Some(trait_did) = tcx.trait_of_item(def_id) {
+        println!("trait: {}", tcx.is_diagnostic_item(sym::VecAllocator, trait_did));
+    }
+
    if feature == sym::allocator_api {
```

Compiler Output:

```
type: false
```

Reply:
That's surprising, I would have expected that to work. I don't mind if you would want to revert to checking the parent is a `Vec` and land this change with that implementation, but you could also look into what `is_diagnostic_item` is doing, we could probably get it to work in this case.

Reply:
@davidtwco 
Do you mean `is_diagnostic_item` doesn't work correctly? If so, I would like to inspect `is_diagnostic_item` and make it work as we expected rather than reverting.

Reply:
I think if we could make `is_diagnostic_item` return `true` for this case (which seems like a reasonable thing for it to do, if other attributes, like those for stability, work in this position), then that would probably be a neater solution, so feel free to do some investigation.

Reply:
@davidtwco 
Thank you for the clarification. I would work on it, but I will split this PR into two if it becomes much complicated.

Reply:
@davidtwco 
Since I do not have enough time anymore to work on this problem, I would like to finish this pull request and open a new pull request when I can.

> I don't mind if you would want to revert to checking the parent is a `Vec` and land this change with that implementation

Also, just to confirm, are you sure you want me to revert it as well? I mean, I thought you already confirmed that we apply this suggestion for all types, including `Vec`. I think the current change satisfies this PR's aim.

If you would like me to revert to restrict the suggestion for only `Vec`, I will revert it.

Reply:
It's just that the current version only checks if the feature is `allocator_api` - I'm not sure if there are other cases where `allocator_api`  might produce a stability error that isn't related to generic parameters where this suggestion wouldn't be applicable. Ideally we'd at least additionally check that `def_id` was a generic parameter or something like that.

Reply:
@davidtwco 
I see. Thank you.
I've reverted to check the parent is `Vec`. I'd like to implement your idea on another PR additionally.